File: app/src/main/java/com/aandssoftware/aandsinventory/listing/CustomerListAdapter.java
package com.aandssoftware.aandsinventory.listing;

import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import androidx.appcompat.app.AlertDialog;
import androidx.cardview.widget.CardView;
import androidx.core.app.ActivityCompat;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import com.aandssoftware.aandsinventory.R;
import com.aandssoftware.aandsinventory.common.Utils;
import com.aandssoftware.aandsinventory.database.RealmManager;
import com.aandssoftware.aandsinventory.models.CallbackRealmObject;
import com.aandssoftware.aandsinventory.models.CustomerModel;
import com.aandssoftware.aandsinventory.ui.ListingActivity;
import com.aandssoftware.aandsinventory.ui.OrderListActivity;
import com.aandssoftware.aandsinventory.ui.adapters.BaseRealmAdapter.BaseViewHolder;
import io.realm.RealmObject;
import io.realm.RealmResults;

public class CustomerListAdapter implements ListingOperations {
  
  private ListingActivity activity;
  public static final String CUSTOMER_ID = "customer_id";
  
  public class CustomerViewHolder extends BaseViewHolder {
    
    @BindView(R.id.imgCustomerItemEdit)
    ImageView imgCustomerItemEdit;
    @BindView(R.id.imgCustomerItemDelete)
    ImageView imgCustomerItemDelete;
    @BindView(R.id.imgCustomerItemLogo)
    ImageView imgCustomerItemLogo;
    @BindView(R.id.tvCustomerName)
    TextView tvCustomerName;
    @BindView(R.id.tvContactNameAndNumber)
    TextView tvContactNameAndNumber;
    @BindView(R.id.tvCustomerGstNumber)
    TextView tvCustomerGstNumber;
    @BindView(R.id.card_view)
    CardView cardView;
    
    @OnClick(R.id.imgCustomerItemDelete)
    public void onDeleteClick() {
      deleteInventory((CustomerModel) itemView.getTag(), itemView.getContext());
    }
    
    @OnClick(R.id.imgCustomerItemEdit)
    public void onEditClick() {
      showAddCustomerFragment((CustomerModel) itemView.getTag(), true);
    }
    
    public CustomerViewHolder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);
    }
  }
  
  public CustomerListAdapter(ListingActivity activity) {
    this.activity = activity;
  }
  
  @Override
  public int getActivityLayoutId() {
    return R.layout.activity_listing;
  }
  
  
  @Override
  public BaseViewHolder getBaseViewHolder(ViewGroup viewGroup, int i) {
    return new CustomerViewHolder(LayoutInflater.from(viewGroup.getContext())
        .inflate(R.layout.customer_item, viewGroup, false));
  }
  
  @Override
  public void onBindSearchViewHolder(BaseViewHolder baseHolder, int position, RealmObject item) {
    CustomerViewHolder holder = (CustomerViewHolder) baseHolder;
    CustomerModel mItem = (CustomerModel) item;
    holder.tvCustomerName.setText(mItem.getCustomerName());
    holder.tvContactNameAndNumber.setText(
        mItem.getContactPerson() + " " + mItem
            .getContactPersonNumber());
    holder.tvCustomerGstNumber.setText(mItem.getCustomerGstNumber());
    if (mItem.getImagePath() != null) {
      Bitmap bitmap = BitmapFactory.decodeFile(mItem.getImagePath());
      if (null != holder.imgCustomerItemLogo && null != bitmap) {
        holder.imgCustomerItemLogo.setImageBitmap(bitmap);
      }
    }
    holder.cardView.setOnClickListener(new OnClickListener() {
      @Override
      public void onClick(View v) {
        if (isOrderSelectionCall()) {
          setResult(mItem);
        } else {
          showAddCustomerFragment((CustomerModel) holder.itemView.getTag(), false);
        }
      }
    });
    holder.itemView.setOnClickListener(new OnClickListener() {
      @Override
      public void onClick(View v) {
        if (isOrderSelectionCall()) {
          setResult(mItem);
        } else {
          showAddCustomerFragment(mItem, false);
        }
      }
    });
  }
  
  public void setResult(CustomerModel mItem) {
    Intent intent = new Intent();
    intent.putExtra(CustomerListAdapter.CUSTOMER_ID,
        (mItem.getCustomerID() == 0) ? mItem.getId() : mItem.getCustomerID());
    activity.setResult(ListingActivity.SELECTED, intent);
    activity.finish();
  }
  
  private boolean isOrderSelectionCall() {
    ComponentName callingActivity = activity.getCallingActivity();
    return (callingActivity != null && callingActivity.getClassName()
        .equalsIgnoreCase(OrderListActivity.class.getName()));
  }
  
  @Override
  public String getTitle() {
    return activity.getString(R.string.customer_item);
  }
  
  @Override
  public RealmResults<? extends RealmObject> getResult() {
    return RealmManager.getCustomerDao().getAllCustomers();
  }
  
  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    activity.getMenuInflater().inflate(R.menu.inventory_menu, menu);
    return true;
  }
  
  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case android.R.id.home:
        activity.finish();
        return true;
      case R.id.actionAdd:
        showAddCustomerFragment(null, false);
        return true;
      default:
    }
    return true;
  }
  
  @Override
  public void onBackPressed() {
    activity.finish();
  }
  
  public void showAddCustomerFragment(CustomerModel mItem, boolean update) {
    CustomerFragment fragment = new CustomerFragment();
    fragment.setCallbackRealmObject(new CallbackRealmObject() {
      @Override
      public void getCallBack(boolean result) {
        if (result) {
          Utils.showToast(activity.getString(R.string.customer_save_message), activity);
          activity.reloadAdapter(getResult());
          fragment.dismiss();
        }
      }
    });
    fragment.setCustomerModel(mItem);
    fragment.setShouldUpdate(update);
    fragment.show(activity.getFragmentManager(), getTitle());
  }
  
  public void deleteInventory(CustomerModel customerModel, Context context) {
    AlertDialog.Builder alertDialogBuilderUserInput =
        new AlertDialog.Builder(context);
    alertDialogBuilderUserInput
        .setTitle(context.getString(R.string.remove_customer_item_title))
        .setMessage(context.getString(R.string.remove_customer_item_message))
        .setCancelable(false)
        .setPositiveButton(context.getString(R.string.yes),
            new DialogInterface.OnClickListener() {
              public void onClick(DialogInterface dialogBox, int id) {
                RealmManager.getCustomerDao().removeCompany(customerModel);
                activity.reloadAdapter(getResult());
              }
            })
        .setNegativeButton(
            context.getString(R.string.no),
            new DialogInterface.OnClickListener() {
              public void onClick(DialogInterface dialogBox, int id) {
                dialogBox.cancel();
              }
            });
    
    final AlertDialog alertDialog = alertDialogBuilderUserInput.create();
    alertDialog.show();
  }
}

File: app/src/main/java/com/aandssoftware/aandsinventory/listing/CustomerListAdapter.kt
package com.aandssoftware.aandsinventory.listing

import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.net.Uri
import android.view.*
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.cardview.widget.CardView
import com.aandssoftware.aandsinventory.R
import com.aandssoftware.aandsinventory.common.Navigator
import com.aandssoftware.aandsinventory.firebase.FirebaseUtil
import com.aandssoftware.aandsinventory.models.CustomerModel
import com.aandssoftware.aandsinventory.models.ViewMode
import com.aandssoftware.aandsinventory.ui.activity.ListingActivity
import com.aandssoftware.aandsinventory.ui.activity.OrderListActivity
import com.aandssoftware.aandsinventory.ui.adapters.BaseAdapter.BaseViewHolder
import com.aandssoftware.aandsinventory.utilities.AppConstants
import com.aandssoftware.aandsinventory.utilities.AppConstants.Companion.RELOAD_LIST_RESULT_CODE
import com.bumptech.glide.Glide
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener
import kotlinx.android.synthetic.main.activity_add_customer.*
import kotlinx.android.synthetic.main.customer_item.view.*
import java.io.Serializable
import java.util.*

class CustomerListAdapter(private val activity: ListingActivity) : ListingOperations {

    companion object {
        val CUSTOMER_ID = "customer_id"
    }

    private val isOrderSelectionCall: Boolean
        get() {
            val callingActivity = activity.callingActivity
            return callingActivity != null && callingActivity.className
                    .equals(OrderListActivity::class.java.name, ignoreCase = true)
        }

    inner class CustomerViewHolder(itemView: View) : BaseViewHolder(itemView) {

        var imgCustomerItemEdit: ImageView = itemView.imgCustomerItemEdit
        var imgCustomerItemDelete: ImageView = itemView.imgCustomerItemDelete
        var imgCustomerItemLogo: ImageView = itemView.imgCustomerItemLogo
        var tvCustomerName: TextView = itemView.tvCustomerName
        var tvContactNameAndNumber: TextView = itemView.tvContactNameAndNumber
        var tvCustomerGstNumber: TextView = itemView.tvCustomerGstNumber
        var cardView: CardView = itemView.cardView

        init {
            imgCustomerItemDelete.setOnClickListener {
                var pos: Int = itemView.getTag(R.string.tag) as Int
                deleteCustomer(itemView.tag as CustomerModel, itemView.context, pos)
            }
            imgCustomerItemEdit.setOnClickListener {
                openCustomerScreen((itemView.tag as CustomerModel).id,
                        (itemView.tag as CustomerModel).customerID, ViewMode.UPDATE.ordinal)
            }
        }
    }

    override fun getActivityLayoutId(): Int {
        return R.layout.activity_listing
    }


    override fun getBaseViewHolder(viewGroup: ViewGroup, i: Int): BaseViewHolder {
        return CustomerViewHolder(LayoutInflater.from(viewGroup.context)
                .inflate(R.layout.customer_item, viewGroup, false))
    }

    override fun onBindSearchViewHolder(baseHolder: BaseViewHolder, position: Int, item: Serializable) {
        val holder = baseHolder as CustomerViewHolder
        val mItem = item as CustomerModel
        holder.tvCustomerName.text = mItem.customerName
        holder.tvContactNameAndNumber.text = mItem.contactPerson + " " + mItem
                .contactPersonNumber
        holder.tvCustomerGstNumber.text = mItem.customerGstNumber
        if (mItem.imagePath != null) {
            if (mItem.imagePath.contains(AppConstants.HTTP, ignoreCase = true)) {
                var uri: Uri = Uri.parse(mItem.imagePath)
                Glide.with(activity)
                        .load(uri)
                        .placeholder(android.R.drawable.ic_menu_gallery)
                        .crossFade()
                        .into(holder.imgCustomerItemLogo)
            } else {
                val bitmap = BitmapFactory.decodeFile(mItem.imagePath)
                if (null != holder.imgCustomerItemLogo && null != bitmap) {
                    holder.imgCustomerItemLogo.setImageBitmap(bitmap)
                }
            }
        }
        holder.cardView.setOnClickListener { v ->
            if (isOrderSelectionCall) {
                setResult(mItem)
            } else {
                openCustomerScreen(item.id, item.customerID,
                        ViewMode.VIEW_ONLY.ordinal)
            }
        }
    }

    private fun setResult(mItem: CustomerModel) {
        val intent = Intent()
        intent.putExtra(CustomerListAdapter.CUSTOMER_ID,
                if (!mItem.id.isEmpty()) mItem.id else mItem.customerID)
        activity.setResult(ListingActivity.SELECTED, intent)
        activity.finish()
    }

    override fun getTitle(): String {
        return activity.getString(R.string.customer_item)
    }

    override fun getResult() {
        activity.showProgressBar()
        FirebaseUtil.getInstance().getCustomerDao().getAllCustomers(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val result = ArrayList<CustomerModel>()
                if (null != dataSnapshot.value) {
                    for (children in dataSnapshot.children) {
                        val model = children.getValue(CustomerModel::class.java)
                        if (null != model) {
                            result.add(model)
                        }
                    }
                    if (!result.isEmpty()) {
                        activity.loadData(result)
                    }
                }
                activity.dismissProgressBar()
            }

            override fun onCancelled(databaseError: DatabaseError) {
                activity.dismissProgressBar()
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        activity.menuInflater.inflate(R.menu.inventory_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                activity.finish()
                return true
            }
            R.id.actionAdd -> {
                openCustomerScreen(AppConstants.EMPTY_STRING, AppConstants.EMPTY_STRING,
                        ViewMode.ADD.ordinal)
                return true
            }
        }
        return true
    }

    override fun onBackPressed() {
        activity.finish()
    }

    fun openCustomerScreen(customerId: String, numericcCustomerId: String, viewMode: Int) {
        Navigator.openCustomerScreen(activity, customerId, numericcCustomerId, viewMode, title)
    }

    fun deleteCustomer(customerModel: CustomerModel, context: Context, pos: Int) {
        val alertDialogBuilderUserInput = AlertDialog.Builder(context)
        alertDialogBuilderUserInput
                .setTitle(context.getString(R.string.remove_customer_item_title))
                .setMessage(context.getString(R.string.remove_customer_item_message))
                .setCancelable(false)
                .setPositiveButton(context.getString(R.string.yes)
                ) { _, _ ->
                    FirebaseUtil.getInstance().getCustomerDao().removeCustomer(customerModel,
                            DatabaseReference.CompletionListener { databaseError, _ ->
                                if (databaseError == null) {
                                    activity.removeAt(pos)

                                }
                            })
                }
                .setNegativeButton(
                        context.getString(R.string.no)
                ) { dialogBox, _ -> dialogBox.cancel() }

        val alertDialog = alertDialogBuilderUserInput.create()
        alertDialog.show()
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            AppConstants.GET_CUSTOMER_UPDATE_REQUEST_CODE // when we open add customer screen
            -> when (resultCode) {
                RELOAD_LIST_RESULT_CODE ->
                    getResult()
            }
        }
    }

}
